### PR TITLE
Lock upload during finalization

### PR DESCRIPTION
**Pull request by TheOneRing**
| Client| Server| State|
|---|---|---|
| A | A |Same version on both sides
| B | A | Versions differ, client starts upload
| B | A | During the upload, B gets locked
| B | B | B is uploaded, Local B is locked so we don't enter the state to the db
| C | B | The local db was not updated, we don't know about B on the server, local file gets updated and becomes C
| C | B | We have C locally and B on the server -> clash

### CODE REVIEW

File: src/common/utility_win.cpp [hunk 1]
```diff
@@ -319,4 +319,31 @@ Utility::NtfsPermissionLookupRAII::~NtfsPermissionLookupRAII()
     qt_ntfs_permission_lookup--;
 }
 
+
+Utility::Handle::Handle(HANDLE h, std::function<void(HANDLE)> &&close)
+    : _handle(h)
```
Comment (dragotin) [inline src/common/utility_win.cpp:324]: is the close function used anywhere so far? If not I'd rather not have it. You're only using the default, right?

**Review by dragotin — COMMENTED** [review]
This patch means that any call to `isFileLocked()` creates a lock on the file, which is released automatically. That function is called often, are you sure there are no side effects?

File: src/common/utility_win.cpp (hunk 1 above)
Comment (TheOneRing) [reply src/common/utility_win.cpp:324]: In the vfs plugin, the handle class is just moved to the main repo.

**Comment by TheOneRing** [comment]
> This patch means that any call to `isFileLocked()` creates a lock on the file, which is released automatically. That function is called often, are you sure there are no side effects?

We did the same before, just explicitly called CloseHandle before.

**Review by dragotin — APPROVED** [review]
ok

File: changelog/unreleased/enterprise-5052 [hunk 2]
```diff
@@ -0,0 +1,6 @@
+Bugfix: Don't publish upload if we can't finish the transaction in the client
+
+When a file ges locked during an upload we aborted after the upload finished on the server.
```
Comment (dschmidt) [inline changelog/unreleased/enterprise-5052:3]:
```suggestion
When a file gets locked during an upload we aborted after the upload finished on the server.
```

**Review by dschmidt — APPROVED** [review]
looks reasonable ... but I'm not familiar with part of the code base

**Pull request merged**